include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -450,13 +451,37 @@ namespace internal
     {
       Assert(n_rows == n_cols, ExcNotImplemented());
 
-      auto &&transpose_fill_nm = [](Number *           out,
-                                    const Number *     in,
-                                    const unsigned int n,
-                                    const unsigned int m) {
-        for (unsigned int mm = 0; mm < m; ++mm)
-          for (unsigned int nn = 0; nn < n; ++nn)
-            out[mm + nn * m] = *(in++);
+      std::vector<bool> constrained_dofs(n_rows, false);
+
+      for (unsigned int i = 0; i < n_rows; ++i)
+        {
+          if (mass_matrix[i + i * n_rows] == 0.0)
+            {
+              Assert(derivative_matrix[i + i * n_rows] == 0.0,
+                     ExcInternalError());
+
+              for (unsigned int j = 0; j < n_rows; ++j)
+                {
+                  Assert(mass_matrix[i + j * n_rows] == 0, ExcInternalError());
+                  Assert(mass_matrix[j + i * n_rows] == 0, ExcInternalError());
+                  Assert(derivative_matrix[i + j * n_rows] == 0,
+                         ExcInternalError());
+                  Assert(derivative_matrix[j + i * n_rows] == 0,
+                         ExcInternalError());
+                }
+
+              constrained_dofs[i] = true;
+            }
+        }
```

Comment:
@kronbichler I have refactored the setup of the constrained mask. What do you think?